lib/pgp.ml

```
@@ -0,0 +1,602 @@
+      let offset =
+        match public_key.version with
+        | 3 -> Ok 8
+        | 4 -> Ok 6
+        | _ -> Error "Bad version of Public key packet"
+      in
+      offset >>= fun offset ->
+      let secret_packet = Cstruct.shift cs offset in
+      let convention = Cstruct.get_uint8 secret_packet 0 in
+      decode_convention public_key secret_packet convention
+  end
+
+  module Body = struct
+    type t =
+      | Id of Id.t
+      | Secret_key of Secret_key.t
+      | Public_key of Public_key.t
+      | Signature
+      | Secret_subkey of Secret_key.t
+      | Public_subkey of Public_key.t
+      | Unknown
+    [@@deriving ord, eq, show]
+
+    let decode packet_type packet =
+      match (packet_type : packet_type) with
+      | Id -> Ok (Id (Id.decode packet))
+      | Secret_key -> Secret_key.decode packet >|= fun key -> Secret_key key
+      | Public_key ->
+        Public_key.decode packet >|= fun (_, key) -> Public_key key
+      | Signature -> Ok Unknown
+      | Secret_subkey ->
+        Secret_key.decode packet >|= fun key -> Secret_subkey key
+      | Public_subkey ->
+        Public_key.decode packet >|= fun (_, key) -> Public_subkey key
+      | Session_key
+      | Unknown_packet ->
+        Ok Unknown
+  end
+
+  type t =
+    { header : Header.t
+    ; packet : Body.t }
+  [@@deriving ord, eq, show]
+
+  let decode cs =
+    match Header.decode cs with
+    | Ok (header_length, header) -> (
+      let packet_cs = Cstruct.sub cs header_length header.packet_length in
+      let res = Body.decode header.packet_type packet_cs in
+      match res with
+      | Error _ ->
+        let next_cs = Cstruct.shift cs (header_length + header.packet_length) in
+        Error next_cs
+      (*When a packet can't be parsed, but the header is correct*)
+      | Ok packet ->
+        let next_cs = Cstruct.shift cs (header_length + header.packet_length) in
+        Ok (next_cs, {header; packet}) )
+    | Error Fatal -> Error Cstruct.empty
```

Using a special value implicitly like `Cstruct.empty` isn't super clear in my opinion.  What do you think about propagating the `Fatal` error instead?  `decode_rec` would then have to match `Error Fatal` and `Error (Header skip_length)`.

In addition to that, perhaps we could also propagate the error string along with `Fatal` to facilitate debugging?  Same for the non-fatal case (e.g. `Header {skip_length; message}`).